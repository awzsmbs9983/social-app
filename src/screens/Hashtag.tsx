import React from 'react'
import {ListRenderItemInfo, Pressable} from 'react-native'
import {PostView} from '@atproto/api/dist/client/types/app/bsky/feed/defs'
import {msg} from '@lingui/macro'
import {useLingui} from '@lingui/react'
import {useFocusEffect} from '@react-navigation/native'
import {NativeStackScreenProps} from '@react-navigation/native-stack'

import {HITSLOP_10} from 'lib/constants'
import {useInitialNumToRender} from 'lib/hooks/useInitialNumToRender'
import {CommonNavigatorParams} from 'lib/routes/types'
import {shareUrl} from 'lib/sharing'
import {cleanError} from 'lib/strings/errors'
import {sanitizeHandle} from 'lib/strings/handles'
import {enforceLen} from 'lib/strings/helpers'
import {isNative} from 'platform/detection'
import {useSearchPostsQuery} from 'state/queries/search-posts'
import {useSetMinimalShellMode} from 'state/shell'
import {Post} from 'view/com/post/Post'
import {List} from 'view/com/util/List'
import {ViewHeader} from 'view/com/util/ViewHeader'
import {ArrowOutOfBox_Stroke2_Corner0_Rounded} from '#/components/icons/ArrowOutOfBox'
import {
  ListFooter,
  ListHeaderDesktop,
  ListMaybePlaceholder,
} from '#/components/Lists'

const renderItem = ({item}: ListRenderItemInfo<PostView>) => {
  return <Post post={item} />
}

const keyExtractor = (item: PostView, index: number) => {
  return `${item.uri}-${index}`
}

export default function HashtagScreen({
  route,
}: NativeStackScreenProps<CommonNavigatorParams, 'Hashtag'>) {
  const {tag, author} = route.params
  const setMinimalShellMode = useSetMinimalShellMode()
  const {_} = useLingui()
  const initialNumToRender = useInitialNumToRender()
  const [isPTR, setIsPTR] = React.useState(false)

  const fullTag = React.useMemo(() => {
    return `#${decodeURIComponent(tag)}`
  }, [tag])

  const queryParam = React.useMemo(() => {
    if (!author) return fullTag
    return `${fullTag} from:${sanitizeHandle(author)}`
  }, [fullTag, author])

  const headerTitle = React.useMemo(() => {
    return enforceLen(fullTag.toLowerCase(), 24, true, 'middle')
  }, [fullTag])

  const sanitizedAuthor = React.useMemo(() => {
    if (!author) return
    return sanitizeHandle(author)
  }, [author])

  const {
    data,
    isFetchingNextPage,
    isLoading,
    isError,
    error,
    refetch,
    fetchNextPage,
    hasNextPage,
  } = useSearchPostsQuery({query: queryParam})

  const posts = React.useMemo(() => {
    return data?.pages.flatMap(page => page.posts) || []
  }, [data])

  useFocusEffect(
    React.useCallback(() => {
      setMinimalShellMode(false)
    }, [setMinimalShellMode]),
  )

  const onShare = React.useCallback(() => {
    const url = new URL('https://bsky.app')
    url.pathname = `/hashtag/${decodeURIComponent(tag)}`
    if (author) {
      url.searchParams.set('author', author)
    }
    shareUrl(url.toString())
  }, [tag, author])

  const onRefresh = React.useCallback(async () => {
    setIsPTR(true)
    await refetch()
    setIsPTR(false)
  }, [refetch])

  const onEndReached = React.useCallback(() => {
    if (isFetchingNextPage || !hasNextPage || error) return
    fetchNextPage()
  }, [isFetchingNextPage, hasNextPage, error, fetchNextPage])

  return (
    <>
      <ViewHeader
        title={headerTitle}
        subtitle={author ? _(msg`From @${sanitizedAuthor}`) : undefined}
        canGoBack
        renderButton={
          isNative
            ? () => (
                <Pressable
                  accessibilityRole="button"
                  onPress={onShare}
                  hitSlop={HITSLOP_10}>
                  <ArrowOutOfBox_Stroke2_Corner0_Rounded
                    size="lg"
                    onPress={onShare}
                  />
                </Pressable>
              )
            : undefined
        }
      />
      {posts.length < 1 ? (
        <ListMaybePlaceholder
          isLoading={isLoading}
          isError={isError}
          onRetry={refetch}
          emptyType="results"
          emptyMessage={_(msg`We couldn't find any results for that hashtag.`)}
        />
      ) : (
        <List
          data={posts}
          renderItem={renderItem}
          keyExtractor={keyExtractor}
          refreshing={isPTR}
          onRefresh={onRefresh}
          onEndReached={onEndReached}
          onEndReachedThreshold={4}
          // @ts-ignore web only -prf
          desktopFixedHeight
          ListHeaderComponent={
            <ListHeaderDesktop
              title={headerTitle}
              subtitle={author ? _(msg`From @${sanitizedAuthor}`) : undefined}
            />
          }
          ListFooterComponent={
            <ListFooter
              isFetchingNextPage={isFetchingNextPage}
              error={cleanError(error)}
              onRetry={fetchNextPage}
            />
          }
          initialNumToRender={initialNumToRender}
          windowSize={11}
        />
      )}
    </>
  )
}
