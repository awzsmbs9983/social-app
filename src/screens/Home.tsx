import React from 'react'
import {Text, Button, View, SafeAreaView} from 'react-native'
import type {PrimaryTabScreenProps} from '../routes/types'

export const Home = ({navigation}: PrimaryTabScreenProps<'Home'>) => {
  return (
    <SafeAreaView style={{flex: 1}}>
      <View style={{flex: 1}}>
        <Text>Hello world</Text>
        <Button
          title="Go to Jane's profile"
          onPress={() => navigation.navigate('Profile', {name: 'Jane'})}
        />
      </View>
    </SafeAreaView>
  )
}
