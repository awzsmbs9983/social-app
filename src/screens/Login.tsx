import React from 'react'
import {Text, Button, View} from 'react-native'
import {Shell} from '../platform/shell'
import type {RootTabsScreenProps} from '../routes/types'
import {useStores} from '../state'

export function Login({navigation}: RootTabsScreenProps<'Login'>) {
  const store = useStores()
  return (
    <Shell>
      <View style={{justifyContent: 'center', alignItems: 'center'}}>
        <Text style={{fontSize: 20, fontWeight: 'bold'}}>Sign In</Text>
        <Button title="Login" onPress={() => store.session.setAuthed(true)} />
        <Button title="Sign Up" onPress={() => navigation.navigate('Signup')} />
      </View>
    </Shell>
  )
}
