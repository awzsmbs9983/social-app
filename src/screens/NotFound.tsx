import React from 'react'
import {Shell} from '../platform/shell'
import {Text, Button, View} from 'react-native'
import type {RootTabsScreenProps} from '../routes/types'

export const NotFound = ({navigation}: RootTabsScreenProps<'NotFound'>) => {
  return (
    <Shell>
      <View style={{justifyContent: 'center', alignItems: 'center'}}>
        <Text style={{fontSize: 20, fontWeight: 'bold'}}>Page not found</Text>
        <Button title="Home" onPress={() => navigation.navigate('Home')} />
      </View>
    </Shell>
  )
}
