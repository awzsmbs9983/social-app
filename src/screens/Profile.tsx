import React from 'react'
import {Text} from 'react-native'
import type {RootStackScreenProps} from '../routes/types'

export const Profile = ({route}: RootStackScreenProps<'Profile'>) => {
  return <Text>This is {route.params.name}'s profile</Text>
}
