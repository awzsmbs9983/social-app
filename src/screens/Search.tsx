import React from 'react'
import {SafeAreaView, ScrollView, Text, Button, View} from 'react-native'
import type {PrimaryTabScreenProps} from '../routes/types'

export const Search = ({navigation}: PrimaryTabScreenProps<'Search'>) => {
  return (
    <SafeAreaView>
      <ScrollView contentInsetAdjustmentBehavior="automatic">
        <View>
          <Text>Hello world</Text>
          <Button
            title="Go to Jane's profile"
            onPress={() => navigation.navigate('Profile', {name: 'Jane'})}
          />
        </View>
      </ScrollView>
    </SafeAreaView>
  )
}
