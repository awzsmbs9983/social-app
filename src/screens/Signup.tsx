import React from 'react'
import {Text, Button, View, ActivityIndicator} from 'react-native'
import {observer} from 'mobx-react-lite'
import {Shell} from '../platform/shell'
import type {RootTabsScreenProps} from '../routes/types'
import {useStores} from '../state'

export const Signup = observer(
  ({navigation}: RootTabsScreenProps<'Signup'>) => {
    const store = useStores()
    return (
      <Shell>
        <View style={{justifyContent: 'center', alignItems: 'center'}}>
          <Text style={{fontSize: 20, fontWeight: 'bold'}}>Create Account</Text>
          {store.session.uiError ?? <Text>{store.session.uiError}</Text>}
          {store.session.uiState === 'idle' ? (
            <>
              <Button
                title="Create new account"
                onPress={() =>
                  store.session.createTestAccount('http://localhost:1986')
                }
              />
              <Button
                title="Log in to an existing account"
                onPress={() => navigation.navigate('Login')}
              />
            </>
          ) : (
            <ActivityIndicator />
          )}
        </View>
      </Shell>
    )
  },
)
