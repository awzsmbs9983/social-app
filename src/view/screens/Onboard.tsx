import React, {useEffect} from 'react'
import {View} from 'react-native'
import {observer} from 'mobx-react-lite'
import {FeatureExplainer} from '../com/onboard/FeatureExplainer'
import {Follows} from '../com/onboard/Follows'
import {OnboardStage, OnboardStageOrder} from '../../state/models/onboard'
import {useStores} from '../../state'

export const Onboard = observer(() => {
  const store = useStores()

  useEffect(() => {
    // sanity check - bounce out of onboarding if the stage is wrong somehow
    if (!OnboardStageOrder.includes(store.onboard.stage)) {
      store.onboard.stop()
    }
  }, [store.onboard.stage])

  let Com
  if (store.onboard.stage === OnboardStage.Explainers) {
    Com = FeatureExplainer
  } else if (store.onboard.stage === OnboardStage.Follows) {
    Com = Follows
  } else {
    Com = View
  }

  return (
    <View style={{flex: 1, backgroundColor: '#fff'}}>
      <Com />
    </View>
  )
})
