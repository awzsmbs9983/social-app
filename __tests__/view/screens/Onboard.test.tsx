import React from 'react'
import {Onboard} from '../../../src/view/screens/Onboard'
import renderer from 'react-test-renderer'
// import {render} from '../../../../jest/test-utils'

describe('Onboard', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Onboard />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
