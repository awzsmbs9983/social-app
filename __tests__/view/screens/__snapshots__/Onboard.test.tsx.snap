// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Onboard renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "flex": 1,
            "overflow": "hidden",
          },
          undefined,
        ]
      }
    >
      <RNCViewPager
        collapsable={false}
        initialPage={0}
        keyboardDismissMode="on-drag"
        layout={
          Object {
            "height": 0,
            "width": 750,
          }
        }
        layoutDirection="ltr"
        onMoveShouldSetResponderCapture={[Function]}
        onPageScroll={[Function]}
        onPageScrollStateChanged={[Function]}
        onPageSelected={[Function]}
        scrollEnabled={true}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            accessibilityElementsHidden={false}
            importantForAccessibility="auto"
            style={
              Array [
                Object {
                  "flex": 1,
                  "overflow": "hidden",
                },
                Object {
                  "width": 750,
                },
                Array [
                  undefined,
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                ],
              ]
            }
          >
            <View
              style={
                Object {
                  "flex": 1,
                  "paddingHorizontal": 16,
                  "paddingTop": 80,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                    },
                    Array [
                      Object {
                        "fontSize": 42,
                        "fontWeight": "bold",
                        "marginBottom": 16,
                        "textAlign": "center",
                      },
                      Object {
                        "fontWeight": "400",
                      },
                      Object {
                        "lineHeight": 60,
                        "paddingBottom": 50,
                        "paddingTop": 50,
                      },
                    ],
                  ]
                }
              >
                Welcome to 
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                      },
                      Array [
                        Object {
                          "fontWeight": "bold",
                        },
                        Object {
                          "color": "#0085ff",
                        },
                        Object {
                          "fontSize": 56,
                        },
                      ],
                    ]
                  }
                >
                  Bluesky
                </Text>
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                    },
                    Array [
                      Object {
                        "fontSize": 18,
                        "marginBottom": 16,
                        "textAlign": "center",
                      },
                      Object {
                        "fontSize": 24,
                      },
                    ],
                  ]
                }
              >
                Let's do a quick tour through the new features.
              </Text>
            </View>
          </View>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            accessibilityElementsHidden={true}
            importantForAccessibility="no-hide-descendants"
            style={
              Array [
                Object {
                  "flex": 1,
                  "overflow": "hidden",
                },
                Object {
                  "width": 750,
                },
                Array [
                  undefined,
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                ],
              ]
            }
          />
        </View>
      </RNCViewPager>
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        />
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "opacity": 1,
              "padding": 16,
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            °
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "opacity": 1,
              "padding": 16,
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            °
          </Text>
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "paddingBottom": 24,
          "paddingHorizontal": 32,
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
              },
              Array [
                Object {
                  "color": "#0085ff",
                },
                Object {
                  "fontSize": 18,
                },
              ],
            ]
          }
        >
          Skip
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      />
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
              },
              Array [
                Object {
                  "color": "#0085ff",
                },
                Object {
                  "fontSize": 18,
                },
              ],
            ]
          }
        >
          Next
        </Text>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
